### ISSUE TITLE
Fails to warn if test subject is hash element

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#15541]
### Before submitting

- [x] I have searched the existing issues and this hasn't been reported yet.
- [x] I am running the latest stable version of RuboCop.

### Issue type

False negative

### Cop name

Lint/OutOfRangeRegexpRef

### Steps to reproduce the problem

```ruby
foo = -> { it[:bar].match?(/(b)/) && $1 }

pp foo.call({ bar: 'a' })
pp foo.call({ bar: 'b' })
```

### Expected behavior

`$1 is out of range`

### Actual behavior

No cop.

Cop shows up if removing `[:bar]`. Not sure if there might be some non-obvious reason for the discrepancy. 

### RuboCop configuration

```yaml
/dev/null
```

### RuboCop version

```shell
System Version: macOS 26.5.2 (25F84)
Model Identifier: MacBookAir10,1
ruby 4.0.6 (2026-07-14 revision 03b6d3f889) +PRISM [arm64-darwin25]
rubocop     2026-08-02 3b9755f3d
```
